tests/test_byteutils.nim

```diff
@@ -65,6 +67,17 @@ suite "Byte utils":
     expect(ValueError): discard hexToByteArray[1]("")
     expect(ValueError): discard hexToByteArray[1]("1")
 
+  test "hexToByteArray: Invalid value":
+    expect(ValueError):
+      var buffer: array[1, byte]
+      hexToByteArray("0x", buffer)
+
+    expect(ValueError):
+      var buffer: distinct seq[byte]
+      hexToByteArray("0x", (openArray[byte])buffer)
```

Comment:
So I think the problem is here. `openArray[byte].len` is `0` and `openArray.high` returns `-1` I think from your other comment.

It's a little hard to test exactly what it is without transforming everything from `func` to `proc` so I can stick some debugging in there. 🤔

Reply:
https://nim-lang.org/docs/system.html#debugEcho%2Cvarargs%5Btyped%2C%5D

Reply:
> So I think the problem is here

yes, this would be a problem - in your test, the output buffer is empty, so there's nothing to write to - when an empty output buffer is passed in, we should probably allow this and not raise anything (no valueerror and no assertion)

Reply:
@arnetheduck However in this case we have an invalid hex input `0x`. It's the only case I've found where this bug occurs but I can check some negative cases also.

@tersec 🔥 Thank you!

Reply:
> However in this case we have an invalid hex input 0x

I'd consider this a valid serialization of an empty byte sequence with `0x` prefix - ie the caller passed in a 0-byte buffer, so they expect that the hex string is a valid zero-byte representation of the empty buffer, which means either an empty string or just a `0x` prefix

Reply:
@arnetheduck since `openArray` cannot be resized, this seems like it should be some kind of error 🤔

Reply:
There are two flavours of `hexStringToArray` functions in this module. The `*Strict` versions require the input string size to match exactly the size of the the output array, while the non-strict versions require only that the input string has enough characters to fill the output array. Both don't resize the output array (as this is not possible with `openArray` as you pointed out yourself).

@arnetheduck's point is that if the user asked for an empty string, it's fine if the non-strict flavour accepts that without an error. The strict flavour is likely to error out unless the input string is also empty (i.e. "" or "0x").

Reply:
@zah it seems though like I can ask for an actual value to be filled into a mistakenly empty array:

```nim
  test "valid hex with empty array":
    var buffer: seq[byte]
    hexToByteArray("0x123", openArray[char](buffer))
```

Which will end up giving the user a weird and hard to find bug in their code. Should this not be an assertion error?